File: libhttpsv/Class1.vb
Imports System.Net
Imports System.Text
Imports System.IO
Imports System.Threading
Imports System.Text.RegularExpressions


Public Class Eventos
    Delegate Sub ReceivedNewRequest(ByVal Request As HttpListenerRequest)
    Delegate Sub ReceivedNewParameter(ByVal Key As String, Value As String)
    Delegate Sub ChangeStatusServer(ByVal Status As StatusServer)
End Class

Public Enum StatusServer As Integer
    ServerOff = 0
    ServerOn = 1
    ServerLoading = 2
End Enum

Public Class HTTPServer
    Private listener As New HttpListener
    Private tStatus As Integer
    Private tshowconsole As Boolean
    Private EndGetContext As HttpListenerContext = Nothing
    Private Paraments As New Dictionary(Of String, String)
    Private tContextFolder As String
    Private Property ContextFiles As New List(Of String)

    Event ReceivedNewRequest As Eventos.ReceivedNewRequest
    Event ReceivedNewParameter As Eventos.ReceivedNewParameter
    Event ChangeStatusServer As Eventos.ChangeStatusServer

    ''' <summary>
    ''' No 'ContextFolder' você pode adicionar a localização de um directório, se existir o arquivo index.html ele será adicionado no Context para a localização raiz do dominio.
    ''' </summary>
    Public Property ContextFolder As String
        Get
            Return tContextFolder
        End Get
        Set(value As String)
            ContextFiles.Clear()
            ContextFiles.AddRange(Directory.GetFiles(value, "*.*", SearchOption.AllDirectories))
            tContextFolder = value
        End Set
    End Property
    ''' <summary>
    ''' O 'Context' irá mostrar a página inicial do site, você pode adicionar códigos HTML, CSS e JavaScript, essa váriavel não vai funcionar se o 'ContextFolder' tiver um valor.
    ''' </summary>
    Public Property Context As String
    ''' <summary>
    ''' O ContextEncoding irá codificar a String definido na variável Context, se não definir um encoding, por padrão a codificação será UTF8.
    ''' </summary>
    Public Property ContextEncoding As Encoding = Encoding.UTF8
    ''' <summary>
    ''' Ao definir valor True nessa variável e caso seu software esteja sendo executado por aplicativo console, você verá todos os logs do HTTPServer.
    ''' </summary>
    Public Property ProcessLogs As Boolean
        Get
            Return tshowconsole
        End Get
        Set(value As Boolean)
            tshowconsole = value
        End Set
    End Property
    ''' <summary>
    ''' Aqui você pode pegar os parâmetros de uma URL, por exemplo: no acesso http://127.0.0.1:8030/?download=ArquivoPDF, ao digitar GetParameter("download"), o valor 'ArquivoPDF' será retornado.
    ''' </summary>
    Public ReadOnly Property GetParameter(ByVal Key As String)
        Get
            If Paraments.ContainsKey(Key) Then
                Key = Paraments(Key)
            Else
                Key = Nothing
            End If
            Return WebUtility.UrlDecode(Key)
        End Get
    End Property

    Public ReadOnly Property Request As HttpListenerRequest
        Get
            Return EndGetContext.Request
        End Get
    End Property

    Public ReadOnly Property Response As HttpListenerResponse
        Get
            Return EndGetContext.Response
        End Get
    End Property

    Public ReadOnly Property User As System.Security.Principal.IPrincipal
        Get
            Return EndGetContext.User
        End Get
    End Property

    ''' <summary>
    ''' Status do servidor HTTP.
    ''' </summary>
    Public ReadOnly Property Status As StatusServer
        Get
            Return tStatus
        End Get
    End Property

    ''' <summary>
    ''' HostStart irá ligar o servidor HTTP em um IP Local com uma porta definido.
    ''' </summary>
    Public Sub HostStart(ByVal Port As Integer)
        If listener.IsListening Then
            CWrite("O servidor já está ligado!")
        Else
            CWrite("Iniciando servidor...")
            tStatus = StatusServer.ServerLoading
            RaiseEvent ChangeStatusServer(tStatus)
            Dim strHostName = System.Net.Dns.GetHostName()
            Dim strIPAddress = System.Net.Dns.GetHostByName(strHostName).AddressList(0).ToString()
            listener.Prefixes.Add("http://" & strIPAddress & ":" & Port & "/")
            listener.Start()
            CWrite("Servidor iniciado com sucesso em: http://" & strIPAddress & ":" & Port & "/")
            tStatus = StatusServer.ServerOn
            RaiseEvent ChangeStatusServer(tStatus)
            listener.BeginGetContext(AddressOf RequestHandler, Interlocked.Increment(0))
        End If
    End Sub

    ''' <summary>
    ''' HostStop irá desligar o servidor HTTP.
    ''' </summary>
    Public Sub HostStop()
        If listener.IsListening Then
            CWrite("Parando servidor...")
            tStatus = StatusServer.ServerLoading
            RaiseEvent ChangeStatusServer(tStatus)
            listener.Stop()
            listener.Abort()
            CWrite("Servidor desligado com sucesso!")
            tStatus = StatusServer.ServerOff
            RaiseEvent ChangeStatusServer(tStatus)
        Else
            CWrite("O servidor já está parado!")
        End If
    End Sub

    Private Sub RequestHandler(ByVal result As IAsyncResult)
        Try
            EndGetContext = listener.EndGetContext(result)
            CWrite("Nova acesso solicitado.   (URL: " & EndGetContext.Request.Url.ToString & ")")
            Paraments.Clear()

            Dim matches As MatchCollection = Regex.Matches(EndGetContext.Request.RawUrl, "(\?|\&)([^=]+)\=([^&]+)")
            For Each m As Match In matches
                If Not Paraments.ContainsKey(m.Groups(2).Value) Then
                    CWrite("Novo parâmetro recebido.  (KEY: " & m.Groups(2).Value & ", VALUE: " & m.Groups(3).Value & ")")
                    Paraments.Add(m.Groups(2).Value, m.Groups(3).Value)
                    RaiseEvent ReceivedNewParameter(m.Groups(2).Value, m.Groups(3).Value)
                End If
            Next

            Dim t As New Thread(Sub() Connection())
            t.Start()

            RaiseEvent ReceivedNewRequest(EndGetContext.Request)
            If Status = StatusServer.ServerOn Then
                listener.BeginGetContext(AddressOf RequestHandler, Interlocked.Increment(0))
            End If
        Catch ex As Exception

        End Try
    End Sub

    Private Sub Connection()
        If Directory.Exists(ContextFolder) Then
            If File.Exists(ContextFolder + "\index.html") And Request.Url.AbsolutePath = "/" Then
                Dim buffer() As Byte = ContextEncoding.GetBytes(File.ReadAllText(ContextFolder + "\index.html"))
                Response.OutputStream.Write(buffer, 0, buffer.Length)
                Response.Close()
            Else
                Context = Nothing
                Dim url As New Uri(ContextFolder + Request.Url.AbsolutePath)
                If File.Exists(url.LocalPath) Then
                    Try
                        Dim FileStream = File.OpenRead(url.LocalPath)
                        EndGetContext.Response.ContentLength64 = FileStream.Length
                        FileStream.CopyTo(Response.OutputStream)
                        Response.Close()
                    Catch ex As Exception

                    End Try
                Else
                    For i As Integer = 0 To ContextFiles.Count - 1
                        Context += "<a href='" & ContextFiles(i).Replace(ContextFolder, "") & "'>" & ContextFiles(i).Replace(ContextFolder, "") & "</a></br>"
                    Next i
                    Dim buffer() As Byte = ContextEncoding.GetBytes(Context)
                    Response.OutputStream.Write(buffer, 0, buffer.Length)
                    Response.Close()
                End If
            End If
        Else
            If Not Context = Nothing Then
                Dim buffer() As Byte = ContextEncoding.GetBytes(Context)
                Response.OutputStream.Write(buffer, 0, buffer.Length)
                Response.Close()
            End If
        End If
    End Sub

    Private Sub CWrite(ByVal text As String)
        If tshowconsole Then
            Console.ForegroundColor = ConsoleColor.Green
            Console.WriteLine("[HTTPServer] " & text)
        End If
    End Sub
End Class

File: Class1.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Net;
using System.Text;
using System.Text.RegularExpressions;
using System.Threading;

namespace libhttpsv{
    public partial class Eventos: EventArgs{
        public delegate void ReceivedNewRequest(HttpListenerRequest Request);
        public delegate void ReceivedNewParameter(string Key, string Value);
        public delegate void ChangeStatusServer(StatusServer Status);
    }

    public enum StatusServer : int{
        ServerOff = 0,
        ServerOn = 1,
        ServerLoading = 2
    }

    public class HTTPServer{
        private Dictionary<int, string> RequestPages = new Dictionary<int, string>(){
            {0, "<html><head><title>Libhttpsv</title><meta charset='UTF8'/></head><body><h1>Libhttpsv</h1>"},
            {1, "<hr/>Libhttpsv 1.0.3.0 - Criado por Trevias Xk</body></html>"},
            {404, "<html><head><title>404 Não encontrado</title><meta charset='UTF8'/></head><body><h1>Não encontrado</h1><p>A url solicitado não foi entrado nesse servidor.</p><hr/>Libhttpsv 1.0.3.0 - Criado por Trevias Xk</body></html>"}
        };
        private HttpListener listener = new HttpListener();
        private int tStatus;
        private bool tshowconsole;
        private bool tlockconsole;
        private HttpListenerContext EndGetContext = default;
        private Dictionary<string, string> Paraments = new Dictionary<string, string>();
        private string tContextFolder;
        private List<string> ContextFiles { get; set; } = new List<string>();
        public event Eventos.ReceivedNewRequest ReceivedNewRequest;
        public event Eventos.ReceivedNewParameter ReceivedNewParameter;
        public event Eventos.ChangeStatusServer ChangeStatusServer;

        /// <summary>
        /// No 'ContextFolder' você pode adicionar a localização de um directório, se existir o arquivo index.html ele será adicionado no Context para a localização raiz do dominio.
        /// </summary>
        public string ContextFolder{get {return tContextFolder;} set {ContextFiles.Clear(); ContextFiles.AddRange(Directory.GetFiles(value, "*.*", SearchOption.AllDirectories)); tContextFolder = value;}
        }
        /// <summary>
        /// O 'Context' irá mostrar a página inicial do site, você pode adicionar códigos HTML, CSS e JavaScript, essa váriavel não vai funcionar se o 'ContextFolder' tiver um valor.
        /// </summary>
        public string Context { get; set; }
        /// <summary>
        /// O ContextEncoding irá codificar a String definido na variável Context, se não definir um encoding, por padrão a codificação será UTF8.
        /// </summary>
        public Encoding ContextEncoding { get; set; } = Encoding.UTF8;
        /// <summary>
        /// Ao definir valor True nessa variável e caso seu software esteja sendo executado por aplicativo console, você verá todos os logs do HTTPServer.
        /// </summary>
        public bool ProcessLogs{get {return tshowconsole;} set {tshowconsole = value;}}
        /// <summary>
        /// Ao definir valor True, o console não será finalizado ao término do código.
        /// </summary>
        public bool ConsoleLock{get {return tlockconsole;} set {tlockconsole = value;}}
        /// <summary>
        /// Aqui você pode pegar os parâmetros de uma URL, por exemplo: no acesso http://127.0.0.1:8030/?download=ArquivoPDF, ao digitar GetParameter("download"), o valor 'ArquivoPDF' será retornado.
        /// </summary>
        public string GetParameter(string Key){
            if (Paraments.ContainsKey(Key)){
                Key = Paraments[Key];
            }
            else {
                Key = null;
            }
            return WebUtility.UrlDecode(Key);
        }

        public HttpListenerRequest Request{get {return EndGetContext.Request;}}

        public HttpListenerResponse Response{get {return EndGetContext.Response;}}

        public System.Security.Principal.IPrincipal User{get {return EndGetContext.User;}}

        /// <summary>
        /// Status do servidor HTTP.
        /// </summary>
        public StatusServer Status{get {return (StatusServer)tStatus;}}

        /// <summary>
        /// HostStart irá ligar o servidor HTTP em um IP e Porta definido na sua rede local.
        /// </summary>
        public void HostStart(int Port, string IP){
            IP = IPAddress.Parse(IP).ToString();
            if (listener.IsListening){
                CWrite("O servidor já está ligado!");
            }
            else {
                try {
                    CWrite("Iniciando servidor...");
                    tStatus = (int)StatusServer.ServerLoading;
                    ChangeStatusServer?.Invoke((StatusServer)tStatus);
                    listener.Prefixes.Add("http://" + IP + ":" + Port + "/");
                    listener.Start();
                    CWrite("Servidor iniciado com sucesso em: http://" + IP + ":" + Port + "/");
                    tStatus = (int)StatusServer.ServerOn;
                    ChangeStatusServer?.Invoke((StatusServer)tStatus);
                    int arglocation = 0;
                    listener.BeginGetContext(RequestHandler, Interlocked.Increment(ref arglocation));
                } catch(UnauthorizedAccessException ex){
                    CWrite("Erro: " + ex.ToString());
                }
            }
            LockConsole();
        }

        private void LockConsole(){
            if(ConsoleLock){
                Console.ReadKey(true);
                LockConsole();
            }
        }

        /// <summary>
        /// HostRestart irá reiniciar o servidor HTTP, é recomendado utilizar esse comando caso o servidor falhe. (Beta)
        /// </summary>

        public void HostRestart(){
            CWrite("Reiniciando servidor...");
            HostStop();
            HostStart();
        }

        /// <summary>
        /// HostStop irá desligar o servidor HTTP, porém comando só irá funcionar se o servidor estiver inicializado por completo.
        /// </summary>
        public void HostStop(){
            if (listener.IsListening){
                CWrite("Parando servidor...");
                tStatus = (int)StatusServer.ServerLoading;
                ChangeStatusServer?.Invoke((StatusServer)tStatus);
                listener.Stop();
                listener.Abort();
                CWrite("Servidor desligado com sucesso!");
                tStatus = (int)StatusServer.ServerOff;
                ChangeStatusServer?.Invoke((StatusServer)tStatus);
            }
            else {
                CWrite("O servidor já está parado!");
            }
        }

        private void RequestHandler(IAsyncResult result){
            try {
                EndGetContext = listener.EndGetContext(result);
                CWrite("Nova acesso solicitado.   (URL: " + EndGetContext.Request.Url.ToString() + ")");
                Paraments.Clear();
                var matches = Regex.Matches(EndGetContext.Request.RawUrl, @"(\?|\&)([^=]+)\=([^&]+)");
                foreach (Match m in matches){
                    if (!Paraments.ContainsKey(m.Groups[2].Value)){
                        CWrite("Novo parâmetro recebido.  (KEY: " + m.Groups[2].Value + ", VALUE: " + m.Groups[3].Value + ")");
                        Paraments.Add(m.Groups[2].Value, m.Groups[3].Value);
                        ReceivedNewParameter?.Invoke(m.Groups[2].Value, m.Groups[3].Value);
                    }
                }



                if (Directory.Exists(ContextFolder)){
                    if(File.Exists(ContextFolder + Request.Url.AbsolutePath + @"\index.html")){
                        byte[] buffer = ContextEncoding.GetBytes(File.ReadAllText(ContextFolder + Request.Url.AbsolutePath + @"\index.html"));
                        Response.OutputStream.Write(buffer, 0, buffer.Length);
                        Response.Close();
                    }
                    else {
                        var url = new Uri(ContextFolder + Request.Url.AbsolutePath);
                        if (File.Exists(url.LocalPath)){
                            var FileStream = File.OpenRead(url.LocalPath);
                            EndGetContext.Response.ContentLength64 = FileStream.Length;
                            FileStream.CopyTo(Response.OutputStream);
                            Response.Close();
                        }
                        else {
                            if(Directory.Exists(ContextFolder + Request.Url.AbsolutePath)){
                                Context = RequestPages[0] + "<p>Lista de arquivos disponíveis nesse diretório.</p><hr/>";
                                for (int i = 0, loopTo = ContextFiles.Count - 1; i <= loopTo; i++){
                                    if(ContextFiles[i].Replace(ContextFolder, "").StartsWith(Request.Url.AbsolutePath.Replace("/",@"\"))){
                                        Context += "<a href='" + ContextFiles[i].Replace(ContextFolder, "") + "'>" + ContextFiles[i].Replace(ContextFolder, "") + "</a></br>";
                                    }
                                }
                                Context += RequestPages[1];
                                byte[] buffer = ContextEncoding.GetBytes(Context);
                                Response.OutputStream.Write(buffer, 0, buffer.Length);
                                Response.Close();
                            }else{
                                byte[] buffer = ContextEncoding.GetBytes(RequestPages[404]);
                                Response.OutputStream.Write(buffer, 0, buffer.Length);
                                Response.Close();
                            }
                        }
                    }
                }
                else if (Context != default){
                    byte[] buffer = ContextEncoding.GetBytes(Context);
                    Response.OutputStream.Write(buffer, 0, buffer.Length);
                    Response.Close();
                }




                ReceivedNewRequest?.Invoke(EndGetContext.Request);
                if (Status == StatusServer.ServerOn){
                    int arglocation = 0;
                    listener.BeginGetContext(RequestHandler, Interlocked.Increment(ref arglocation));
                }
            }
            catch (Exception)
            {
            }
        }

        private void CWrite(string text){
            if (tshowconsole){
                Console.ForegroundColor = ConsoleColor.Green;
                Console.WriteLine("[HTTPServer] " + text);
            }
        }
    }
}
